### PR TITLE
Backend/config file

### LINKED ISSUE #282: [BE] config file
Before this project is done we need a config file to be able to set the following settings:

- [x] Set mongodb url (& mongodb authentication if possible in the timeframe)
- [x] Set elasticsearch url (& mongodb authentication if possible in the timeframe)
- [x] Set listening interface and port (0.0.0.0:5000 for example)
- [x] Set the 'SECRET_KEY' and 'SECURITY_PASSWORD_SALT' variables
- [x] Set a must-have admin account that is created if not existing (mail & password) from the config file

All settings should default to values, if they are not set or if no config file exists, that match our current docker-setup values. This way our containers continue to run even if no config file is found.

The config file should be in any human-readable format like json,yaml,cfg,etc.

The location to read from should be the current folder of the python folder (so `backend` in our case), but an example config file should be placed into the root of the repository so our container does not load.

### PR DESCRIPTION
**Pull request by markwitz**
resolve #282 

done so far

### CODE REVIEW

File: backend/__init__.py [hunk 1]
```diff
@@ -163,9 +196,9 @@ def to_dict(self):
 
 
 class EmailConverter(BaseConverter):
-    regex = r"([a-z0-9!#$%&'*+\/=?^_`{|}~-]+(?:\." +\
-            r"[a-z0-9!#$%&'*+\/=?^_`"r"{|}~-]+)" +\
-            r"*(@|\sat\s)(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?" +\
+    regex = r"([a-z0-9!#$%&'*+\/=?^_`{|}~-]+(?:\." + \
+            r"[a-z0-9!#$%&'*+\/=?^_`"r"{|}~-]+)" + \
+            r"*(@|\sat\s)(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?" + \
```
Comment (henderkes) [inline backend/__init__.py]: please remove the whitespaces

File: backend/__init__.py [hunk 2]
```diff
@@ -27,18 +28,48 @@
...
+        if "flask" in config:
+            default_config["flask"].update(config["flask"])
+        if "mongo_db" in config:
+            default_config["mongo_db"].update(config["mongo_db"])
+        if "elasticsearch" in config:
+            default_config["elasticsearch"].update(config["elasticsearch"])
+        if "administration_user" in config:
+            default_config["administration_user"].update(config["administration_user"])
+
+config = default_config
```
Comment (henderkes) [inline backend/__init__.py:69]: We should put the default_config in a seperate file as well in my opinion.

File: backend/__init__.py [hunk 3]
```diff
@@ -243,10 +276,11 @@ def on_project_deletion():
...
                               '/project/' not in str(x.link) or g.projects.find_one(
-                               {'_id': uuid.UUID(
-                                   str(x.link)[str(x.link).index('/project/') + len('/project/'):]
-                                   )
-                                })]
+                                  {'_id': uuid.UUID(
+                                      str(x.link)[str(x.link).index('/project/') +
+                                                  len('/project/'):]
+                                  )
+                                  })]
```
Comment (henderkes) [inline backend/__init__.py]: Feels harder to read than before

File: tests/api/test_projects.py [hunk 4]
```diff
@@ -276,12 +276,13 @@ def test_inconsistent_delete(self, session, flask_api_url, pytestconfig):
...
         assert delete_response.status_code == 200
 
-    def test_unauthorized_delete(self, flask_api_url):
+    def test_unauthorized_delete(self, flask_api_url, enter_default_user_users):
         """ Tests for 404 when attempting to delete a project when
             logged in as a user.
             Not found because the user should not know about the
             /api/delete endpoint.
         """
+        response = enter_default_user_users
```
Comment (henderkes) [inline tests/api/test_projects.py:285]: why? this is not used anywhere

File: tests/api/test_projects.py [hunk 5]
```diff
@@ -505,9 +506,10 @@ def test_invalid_id(self, session, flask_api_url):
     def test_success(self, session, flask_api_url):
         assert True
 
-    def test_unauthorized_update(self, flask_api_url):
+    def test_unauthorized_update(self, flask_api_url, enter_default_user_users):
         """ Tests for 403 when attempting to update a different users project
         """
+        response = enter_default_user_users
```
Comment (henderkes) [inline tests/api/test_projects.py:513]: again, why?

File: backend/__init__.py (hunk 2 above)
Comment (Drakulix) [reply backend/__init__.py:69]: not necessary

File: backend/__init__.py [hunk 6]
```diff
@@ -27,18 +28,48 @@
...
+default_config["mongo_db"]["secret_key"] = "super-secret"
+default_config["mongo_db"]["hostname"] = "mongodb"
+default_config["mongo_db"]["port"] = 27017
+default_config["mongo_db"]["security_password_hash"] = 'pbkdf2_sha512'
+default_config["mongo_db"]["security_password_salt"] = 'THISISMYOWNSALT'
+default_config["elasticsearch"]["hostname"] = "elasticsearch"
+default_config["elasticsearch"]["port"] = 9200
+default_config["administration_user"]["username"] = "admin"
+default_config["administration_user"]["password"] = "admin"
+default_config["administration_user"]["email"] = "admin@knex.com"
```
Comment (Drakulix) [inline backend/__init__.py]: You could write that a single literal, that would look much nicer.

File: backend/__init__.py (hunk 2 above)
Comment (Drakulix) [inline backend/__init__.py]: not necessary in my opinion.

File: backend/__init__.py (hunk 1 above)
Comment (Drakulix) [reply backend/__init__.py]: Is that a pep8 thing? Seems your editor @markwitz does that all the time?

File: backend/__init__.py (hunk 3 above)
Comment (Drakulix) [reply backend/__init__.py]: Not too important, but @DubbleClick is right.

File: backend/__init__.py [hunk 7]
```diff
@@ -259,18 +293,14 @@ def project_deleted_func():
...
     user_role = USER_DATASTORE.find_or_create_role('user')
     admin_role = USER_DATASTORE.find_or_create_role('admin')
-    userpw = hash_password("user")
-    adminpw = hash_password("admin")
-    try:
-        USER_DATASTORE.create_user(
-            email="user@knex.com", password=userpw, roles=[user_role])
-    except NotUniqueError:
-        pass
+    # userpw = hash_password("user")
```
Comment (Drakulix) [inline backend/__init__.py]: Please remove, if that is useless now.

File: tests/api/test_projects.py (hunk 4 above)
Comment (Drakulix) [reply tests/api/test_projects.py:285]: The test requires the user to exist, this just runs the fixture.

File: tests/api/test_projects.py (hunk 5 above)
Comment (Drakulix) [reply tests/api/test_projects.py:513]: Same. this is good enough for a test.

**Review by Drakulix — CHANGES_REQUESTED** [review]
Good stuff, minor nitpicks.

File: tests/api/test_projects.py (hunk 5 above)
Comment (markwitz) [reply tests/api/test_projects.py:513]: we could check for the response to be correct here but i just assume it is working correctly

File: backend/__init__.py (hunk 6 above)
Comment (markwitz) [reply backend/__init__.py]: i'm at a loss here how do i do that?

Comment (Drakulix) [reply backend/__init__.py]: Looks like json, but is valid python syntax:
```python
default_config = {
    ...,
    'administration_user': {
        'username': 'admin',
        'password': 'admin',
        'email': 'admin@knex.com',
    }
}
```

**Review by henderkes — DISMISSED** [review]
looks good to me now. formatting in on_project_deletion could still be reverted, but it doesn't matter as long as pep8 is happy

**Review by Drakulix — DISMISSED** [review]

**Review by henderkes — APPROVED** [review]

**Pull request merged**